docs/discussion/Multi-Tenancy Support.md

```diff
@@ -0,0 +1,168 @@
+| Cache isolation | easy | hard |
+| Resouce efficiency | low | high |
+| Deployment for new namespace | new deployment | no new deployment |
+| mTLS with GK | manually maintain CA certs | no change |
+
+
+### Deployment per Namespace
+#### Pros
+1. Minimal refacotring on Ratify. Since Ratify is deployed to each required namespace, Ratify itself doesn't need to care too much about the data isolation across namespaces. Instead it can offload multi-tenancy responsibility to the control plane.
+2. Distributed cache service can be deployed per namespace as well to support cache isoaltion easily.
+3. No need to add namespace to the external data request from GK to Ratify. We can keep the same request struct.
+4. Users could easily set up ResourceQuota and LimitRange to assign appropriate resources.
+5. As Ratify pods are isolated by namespace, each team can only access its own logs.
+
+
+#### Cons
+1. Consume more resources. Each Ratify service requires at least one pod, and if HA enabled, Redis cluster is deployed to each namespace as well. In the case of hundreds of tenants, it requires hundreds of Ratify deployment even though there might be few workload being used.
+2. For each Ratify in a namespace, Gatekeeper regards each as a separate external data provider. Therefore, for each admission review to Gatekeeper, GK will send out validation requests to all Ratify services across all namespaces.
+3. More manual work for deployment. Once a new namespace created, a new Ratify deployment is required on it.
+4. How does Ratify trust Gatekeeper(mTLS)? Current Ratify follows the first [approach](https://open-policy-agent.github.io/gatekeeper/website/docs/externaldata/#how-the-external-data-provider-trusts-gatekeeper-mtls) recommended by GK. However, it requires Ratify deployed in the same namespace as GK. Therefore, we have to maintain a cluster-wide, well-known CA cert for GK so that Ratify services in all namespaces could trust it.
+
+#### Implementation Detail
+This option requires minimal code changes to Ratify core workflow as Ratify itself doesn't handle the multi-tenancy scenario. However, we have to make a lot of changes to the helm templates/helmfile, making each Ratify service deployed in each namespace. And all Custom Resources should be namespaced instead of clustered.
+
+### Deployment per Cluster
+#### Pros
+1. More resource efficent. Since there is only one Ratify deployment in the cluster, all tenants share the same Ratify pods and dependency resources. If the load traffic increases, admin could easily scale out the Ratify service accordingly.
+2. GK will work in the current way that sends out one ED request to Ratify for each admission review.
+3. Users only need to deploye once when enabling Ratify feature in the cluster. Afterwards, admins can configure CRs in their own namespaces.
+4. Since GK and Ratify can be deployed to the same namespace, we can enable `cert-controller` of GK to set up mTLS automatically.
+
+#### Cons
```

Comment:
Another con: how do metrics partitioning work? Right now metrics are emitted at the ratify pod level. Would each tenant want their own view of Ratify's perf related to processing requests for their namespace?